### ISSUE TITLE
future/v1: 404's trying to read files from manifest

### ISSUE THREAD

**Issue by t0byman** [brandcom/cakephp-vite#7]
In branch `future/v1` I get 404's trying to read files from manifest with `'forceProductionMode' => true,`. This is my vite.config.js build object:
```
build: {
      manifest: true,
      rollupOptions: {
        input: {
          // Styles
          styles: 'resources/styles/main.css',
          critical: 'resources/styles/critical.css',
          storybook: 'resources/styles/storybook.css',

          // Scripts
          main: 'resources/scripts/main.ts',
        },
        output: {
          entryFileNames: 'assets/[name]-[hash].js',
          chunkFileNames: 'assets/[name]-[hash].js',
          assetFileNames: 'assets/[name]-[hash].[ext]',
        },
      },
      outDir: '../webroot/dist',
      emptyOutDir: true,
    },
```
Which results in this build structure:
![Schermafbeelding_20230121_164448](https://user-images.githubusercontent.com/5767100/213874765-f35973dc-76a8-4def-98e0-c894073a78cd.png)
As you can see the manifest.json gets put inside the dist folder by vite. This is the generated manifest.json:
````
{
  "styles/main.css": {
    "file": "assets/main-e7a50703.css",
    "src": "styles/main.css",
    "isEntry": true
  },
  "styles/storybook.css": {
    "file": "assets/storybook-3b78efcc.css",
    "src": "styles/storybook.css",
    "isEntry": true
  },
  "styles/critical.css": {
    "file": "assets/critical-a9e7c1bb.css",
    "src": "styles/critical.css",
    "isEntry": true
  },
  "scripts/components/t-button/t-button.ce.vue": {
    "file": "assets/t-button.ce-9bcdf683.js",
    "src": "scripts/components/t-button/t-button.ce.vue",
    "isDynamicEntry": true,
    "imports": [
      "scripts/main.ts"
    ]
  },
  "scripts/main.ts": {
    "file": "assets/main-a077a8fb.js",
    "src": "scripts/main.ts",
    "isEntry": true,
    "dynamicImports": [
      "scripts/components/t-button/t-button.ce.vue"
    ]
  }
}
````
Because it says `return '/'` in `ManifestRecord.php` lines 94 and 112 it will return `http://project.test/assets/main-a077a8fb.js` but I want it to return `http://project.test/dist/assets/main-a077a8fb.js`.

Setting `build.outDirectory` to `dist` does nothing because `outDir` isn't used anywhere in `ManifestRecord.php`. I would expect `build.outDirectory` to prefix all url's read from the manifest, can you fix this?

**Comment by passchn** [comment]
@t0byman it should work now in this pre-release: https://github.com/brandcom/cakephp-vite/releases/tag/v1-alpha1 

Also, no need for you to set the `outDirectory` to `dist` because this is the default now. (It was `build` before, but `dist` is the vite default, so the plugin should respect that.)

**Comment by t0byman** [comment]
@passchn ah some nice improvements again! I've tested tag/v1-alpha**2** in production mode again and js files work now! The only thing that's not working yet are css files. I don't even get 404 on css files, it's like they're not being read from the manifest at all. Did you test with a seperate css entry / not imported in js?

**Comment by passchn** [comment]
@t0byman I changed the Api of the `ViteScripts::css()` and `::scripts()` methods. You can't pass file paths anymore and should set them in config like it was/is in `v0.x` Maybe you missed that? 

Could you check with `<?= h($this->fetch('css')) ?>` (escaped html) what the exact output is? And maybe checkout the latest commit.. I'll release a alpha3 now.

**Comment by passchn** [comment]
ok.. I changed that after the release of alpha2.. :D FYI, the latest and greatest:
https://github.com/brandcom/cakephp-vite/compare/v1-alpha2...v1-alpha3

**Comment by t0byman** [comment]
@passchn But how could you ever do code splitting if you have to pass file paths in the config now? Imho you should be able to load specific files from manifest on specific pages. Modern front-end development is not loading every file from manifest on every page, it's loading only the files you need on those pages. E.g. load main.js/ts in layout/default.php but load filters.js/ts only on an overview page. If you agree, maybe revert to passing file paths through the `ViteScripts::css()` and `::scripts()` methods again?

**Comment by passchn** [comment]
Ok, I did not see that use case, but it makes sense now. 

I myself am using css-imports in the main.ts and dynamic imports for modules that I don't need immediately or on every page. So having only one entry is enough and still you have code splitting. 

However, I can make it so that you can pass filters through the $options array. 

I was focusing on making multiple config options possible yesterday night,so that plugins can add files coming from a different manifest file.

**Comment by t0byman** [comment]
I used to import my main.css in my main.ts too but then you can’t use relative paths in your css to your vite public folder like ‘../fonts/testfont-Regular.woff2’. You’ll have to use ‘/dist/fonts/testfont.woff2’ which sucks because then you have to do a ‘vite build’ before you can do a ‘vite dev’ (otherwise the fonts are not in dist yet) and you’ll get a warning like ‘/dist/fonts/testfont.woff2 referenced in C:\repos\myrepo\src\styles\base\fonts.css didn't resolve at build time, it will remain unchanged to be resolved at runtime’ for every public asset used in your css files. Using a separate css entry in vite.config.js fixes all that and you can just use relative paths which work for both vite dev and vite build.

Passing through options is fine then, unless your know a solution with which I can use relative paths in css, import in main.ts and have it work in both vite dev and build without getting that warning.

I won’t be using plugins with a different manifest file though, maybe that’s another use case.

**Comment by passchn** [comment]
The plugin thing is my own use case, that's why I'm interested in doing a heavy makeover in the first place :P 
FYI - I am working on a file selector/pool here on `develop`: https://github.com/brandcom/cakephp-assets/tree/develop 

Wanted to do the frontend-stuff with vite, but as helpers only have one instance, that was not possible. Therefore you ca now pass your own `ViteHelperConfig` to `::script()` and `::css()` 

Anyway, I added a `filter` option for the methods. You can either pass a string, an array or a callback function to filter the `ManifestRecord`s: 

```
$this->ViteScripts->script([
    'filter' => 'main.ts',
]);
```

```
$this->ViteScripts->script([
    'filter' => ['main.ts', 'myEntry.js'],
]);

```

```
$this->ViteScripts->script([
    'filter' => function (\ViteHelper\Utilities\ManifestRecord $record) {
        return $record->matchMany(['myEntry.js', 'main.ts']);
    },
]);
``` 

This is currently only working in production mode. Does this help your with your setup?

**Comment by passchn** [comment]
Changes: https://github.com/brandcom/cakephp-vite/compare/v1-alpha3...v1-alpha4

**Comment by passchn** [comment]
For me, it works with a separate scss-input now.

The only strange thing is that it creates a legacy-js-file from my scss input (I have the legacy plugin installed).

Manifest + Config:  

````
{
  "__commonjsHelpers-legacy-5498c249.js": {
    "file": "assets/_commonjsHelpers-legacy-5498c249.js"
  },
  "____vite-browser-external_commonjs-proxy-legacy-e3df2c0e.js": {
    "file": "assets/___vite-browser-external_commonjs-proxy-legacy-e3df2c0e.js",
    "imports": [
      "__commonjsHelpers-legacy-5498c249.js"
    ]
  },
  "webroot_src/js/index-legacy.js": {
    "file": "assets/index-legacy-9886768d.js",
    "src": "webroot_src/js/index-legacy.js",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts"
    ],
    "dynamicImports": [
      "webroot_src/js/admin/admin-legacy.js"
    ]
  },
  "_js.cookie-legacy-c7672909.js": {
    "file": "assets/js.cookie-legacy-c7672909.js"
  },
  "webroot_src/CodeButton/initCodeButton-legacy.ts": {
    "file": "assets/initCodeButton-legacy-3fc7a49c.js",
    "src": "webroot_src/CodeButton/initCodeButton-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts"
    ]
  },
  "webroot_src/VideoSliders/init-legacy.ts": {
    "file": "assets/init-legacy-c822d183.js",
    "src": "webroot_src/VideoSliders/init-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts"
    ]
  },
  "webroot_src/Alpine/Alpine-legacy.ts": {
    "file": "assets/Alpine-legacy-638f4f89.js",
    "src": "webroot_src/Alpine/Alpine-legacy.ts",
    "isDynamicEntry": true
  },
  "webroot_src/YoutubePlayer/YoutubePlayer-legacy.ts": {
    "file": "assets/YoutubePlayer-legacy-b22053b3.js",
    "src": "webroot_src/YoutubePlayer/YoutubePlayer-legacy.ts",
    "isDynamicEntry": true
  },
  "webroot_src/Chart/Chart-legacy.ts": {
    "file": "assets/Chart-legacy-9fcdad8c.js",
    "src": "webroot_src/Chart/Chart-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts"
    ]
  },
  "webroot_src/js/admin/admin-legacy.js": {
    "file": "assets/admin-legacy-f134f5e0.js",
    "src": "webroot_src/js/admin/admin-legacy.js",
    "isDynamicEntry": true,
    "imports": [
      "__commonjsHelpers-legacy-5498c249.js"
    ]
  },
  "webroot_src/Confetti/Confetti-legacy.ts": {
    "file": "assets/Confetti-legacy-3d717cea.js",
    "src": "webroot_src/Confetti/Confetti-legacy.ts",
    "isDynamicEntry": true
  },
  "webroot_src/main-legacy.ts": {
    "file": "assets/main-legacy-d186e490.js",
    "src": "webroot_src/main-legacy.ts",
    "isEntry": true,
    "dynamicImports": [
      "webroot_src/Cart/init-cart-legacy.ts",
      "webroot_src/Admin/admin-legacy.ts",
      "webroot_src/js/index-legacy.js",
      "webroot_src/VideoPage/video-legacy.ts",
      "webroot_src/Checkout/init-checkout-legacy.ts",
      "webroot_src/CodeButton/initCodeButton-legacy.ts",
      "webroot_src/LoginAndRegistrationPage/init-legacy.ts",
      "webroot_src/VideoSliders/init-legacy.ts",
      "webroot_src/ProductRatings/init-legacy.ts",
      "webroot_src/Alpine/Alpine-legacy.ts",
      "webroot_src/YoutubePlayer/YoutubePlayer-legacy.ts"
    ]
  },
  "webroot_src/Cart/init-cart-legacy.ts": {
    "file": "assets/init-cart-legacy-8c157821.js",
    "src": "webroot_src/Cart/init-cart-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts"
    ]
  },
  "webroot_src/ProductRatings/init-legacy.ts": {
    "file": "assets/init-legacy-9a8789a5.js",
    "src": "webroot_src/ProductRatings/init-legacy.ts",
    "isDynamicEntry": true
  },
  "webroot_src/LoginAndRegistrationPage/init-legacy.ts": {
    "file": "assets/init-legacy-6c4efc27.js",
    "src": "webroot_src/LoginAndRegistrationPage/init-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts",
      "_js.cookie-legacy-c7672909.js"
    ]
  },
  "webroot_src/scss/styles-legacy.scss": {
    "file": "assets/styles-legacy-0168ba53.js",
    "src": "webroot_src/scss/styles-legacy.scss",
    "isEntry": true
  },
  "webroot_src/Checkout/init-checkout-legacy.ts": {
    "file": "assets/init-checkout-legacy-4ad22b48.js",
    "src": "webroot_src/Checkout/init-checkout-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts",
      "_js.cookie-legacy-c7672909.js"
    ],
    "dynamicImports": [
      "webroot_src/Confetti/Confetti-legacy.ts"
    ]
  },
  "webroot_src/VideoPage/video-legacy.ts": {
    "file": "assets/video-legacy-ad864ba6.js",
    "src": "webroot_src/VideoPage/video-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main-legacy.ts",
      "__commonjsHelpers-legacy-5498c249.js",
      "____vite-browser-external_commonjs-proxy-legacy-e3df2c0e.js"
    ]
  },
  "webroot_src/Admin/admin-legacy.ts": {
    "file": "assets/admin-legacy-b29e0524.js",
    "src": "webroot_src/Admin/admin-legacy.ts",
    "isDynamicEntry": true,
    "imports": [
      "__commonjsHelpers-legacy-5498c249.js",
      "____vite-browser-external_commonjs-proxy-legacy-e3df2c0e.js",
      "webroot_src/main-legacy.ts"
    ],
    "dynamicImports": [
      "webroot_src/Chart/Chart-legacy.ts"
    ]
  },
  "vite/legacy-polyfills-legacy": {
    "file": "assets/polyfills-legacy-0302122a.js",
    "src": "vite/legacy-polyfills-legacy",
    "isEntry": true
  },
  "webroot_src/Cart/init-cart.css": {
    "file": "assets/init-cart-ec065b62.css",
    "src": "webroot_src/Cart/init-cart.css"
  },
  "webroot_src/VideoPage/video.css": {
    "file": "assets/video-259ff05f.css",
    "src": "webroot_src/VideoPage/video.css"
  },
  "webroot_src/ProductRatings/init.css": {
    "file": "assets/init-4cb7647d.css",
    "src": "webroot_src/ProductRatings/init.css"
  },
  "webroot_src/LoginAndRegistrationPage/init.css": {
    "file": "assets/init-b86111f0.css",
    "src": "webroot_src/LoginAndRegistrationPage/init.css"
  },
  "webroot_src/main.css": {
    "file": "assets/main-79b2b94e.css",
    "src": "webroot_src/main.css"
  },
  "webroot_src/Admin/admin.css": {
    "file": "assets/admin-a4d9023f.css",
    "src": "webroot_src/Admin/admin.css"
  },
  "webroot_src/Checkout/init-checkout.css": {
    "file": "assets/init-checkout-c1a74a3b.css",
    "src": "webroot_src/Checkout/init-checkout.css"
  },
  "webroot_src/scss/styles.scss": {
    "file": "assets/styles-f43f73cf.css",
    "src": "webroot_src/scss/styles.scss",
    "isEntry": true
  },
  "webroot_src/js/index.js": {
    "file": "assets/index-72135551.js",
    "src": "webroot_src/js/index.js",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts"
    ],
    "dynamicImports": [
      "webroot_src/js/admin/admin.js"
    ]
  },
  "_js.cookie-40e39767.js": {
    "file": "assets/js.cookie-40e39767.js"
  },
  "__commonjsHelpers-87174ba5.js": {
    "file": "assets/_commonjsHelpers-87174ba5.js"
  },
  "webroot_src/CodeButton/initCodeButton.ts": {
    "file": "assets/initCodeButton-8773c5ba.js",
    "src": "webroot_src/CodeButton/initCodeButton.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts"
    ]
  },
  "____vite-browser-external_commonjs-proxy-d2112d3c.js": {
    "file": "assets/___vite-browser-external_commonjs-proxy-d2112d3c.js",
    "imports": [
      "__commonjsHelpers-87174ba5.js"
    ]
  },
  "webroot_src/YoutubePlayer/YoutubePlayer.ts": {
    "file": "assets/YoutubePlayer-b0572043.js",
    "src": "webroot_src/YoutubePlayer/YoutubePlayer.ts",
    "isDynamicEntry": true
  },
  "webroot_src/ProductRatings/init.ts": {
    "file": "assets/init-75d3f0c8.js",
    "src": "webroot_src/ProductRatings/init.ts",
    "isDynamicEntry": true,
    "css": [
      "assets/init-4cb7647d.css"
    ]
  },
  "webroot_src/Confetti/Confetti.ts": {
    "file": "assets/Confetti-c52f9f01.js",
    "src": "webroot_src/Confetti/Confetti.ts",
    "isDynamicEntry": true
  },
  "webroot_src/LoginAndRegistrationPage/init.ts": {
    "file": "assets/init-99209257.js",
    "src": "webroot_src/LoginAndRegistrationPage/init.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts",
      "_js.cookie-40e39767.js"
    ],
    "css": [
      "assets/init-b86111f0.css"
    ]
  },
  "webroot_src/Alpine/Alpine.ts": {
    "file": "assets/Alpine-786fa0d0.js",
    "src": "webroot_src/Alpine/Alpine.ts",
    "isDynamicEntry": true
  },
  "webroot_src/Cart/init-cart.ts": {
    "file": "assets/init-cart-90575c48.js",
    "src": "webroot_src/Cart/init-cart.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts"
    ],
    "css": [
      "assets/init-cart-ec065b62.css"
    ]
  },
  "webroot_src/VideoSliders/init.ts": {
    "file": "assets/init-3f38e00c.js",
    "src": "webroot_src/VideoSliders/init.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts"
    ]
  },
  "webroot_src/Checkout/init-checkout.ts": {
    "file": "assets/init-checkout-ec49448c.js",
    "src": "webroot_src/Checkout/init-checkout.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts",
      "_js.cookie-40e39767.js"
    ],
    "dynamicImports": [
      "webroot_src/Confetti/Confetti.ts"
    ],
    "css": [
      "assets/init-checkout-c1a74a3b.css"
    ]
  },
  "webroot_src/js/admin/admin.js": {
    "file": "assets/admin-b052fa48.js",
    "src": "webroot_src/js/admin/admin.js",
    "isDynamicEntry": true,
    "imports": [
      "__commonjsHelpers-87174ba5.js"
    ]
  },
  "webroot_src/main.ts": {
    "file": "assets/main-8a9f441b.js",
    "src": "webroot_src/main.ts",
    "isEntry": true,
    "dynamicImports": [
      "webroot_src/Cart/init-cart.ts",
      "webroot_src/Admin/admin.ts",
      "webroot_src/js/index.js",
      "webroot_src/VideoPage/video.ts",
      "webroot_src/Checkout/init-checkout.ts",
      "webroot_src/CodeButton/initCodeButton.ts",
      "webroot_src/LoginAndRegistrationPage/init.ts",
      "webroot_src/VideoSliders/init.ts",
      "webroot_src/ProductRatings/init.ts",
      "webroot_src/Alpine/Alpine.ts",
      "webroot_src/YoutubePlayer/YoutubePlayer.ts"
    ],
    "css": [
      "assets/main-79b2b94e.css"
    ]
  },
  "webroot_src/Chart/Chart.ts": {
    "file": "assets/Chart-55f10e5f.js",
    "src": "webroot_src/Chart/Chart.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts"
    ]
  },
  "webroot_src/Admin/admin.ts": {
    "file": "assets/admin-3f915aa3.js",
    "src": "webroot_src/Admin/admin.ts",
    "isDynamicEntry": true,
    "imports": [
      "__commonjsHelpers-87174ba5.js",
      "____vite-browser-external_commonjs-proxy-d2112d3c.js",
      "webroot_src/main.ts"
    ],
    "dynamicImports": [
      "webroot_src/Chart/Chart.ts"
    ],
    "css": [
      "assets/admin-a4d9023f.css"
    ]
  },
  "webroot_src/VideoPage/video.ts": {
    "file": "assets/video-3a28d082.js",
    "src": "webroot_src/VideoPage/video.ts",
    "isDynamicEntry": true,
    "imports": [
      "webroot_src/main.ts",
      "__commonjsHelpers-87174ba5.js",
      "____vite-browser-external_commonjs-proxy-d2112d3c.js"
    ],
    "css": [
      "assets/video-259ff05f.css"
    ]
  }
}

````

````
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';
import legacy from '@vitejs/plugin-legacy';

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        vue(),
        legacy({
            targets: ['defaults', 'not IE 11'],
        }),
    ],
    build: {
        emptyOutDir: true,
        outDir: './webroot/dist',
        manifest: true,
        rollupOptions: {
            input: {
                styles: './webroot_src/scss/styles.scss',
                main: './webroot_src/main.ts',
            },
            output: {
                entryFileNames: 'assets/[name]-[hash].js',
                chunkFileNames: 'assets/[name]-[hash].js',
                assetFileNames: 'assets/[name]-[hash].[ext]',
            }
        },
    },
    server: {
        port: 3000,
        strictPort: true,
        hmr: {
            host: 'localhost',
            protocol: 'ws',
        },
    },
});

````

**Comment by t0byman** [comment]
> Does this help your with your setup?

This definitely works with my setup now, both my css files and js are read from manifest! I also tested dev mode again but now I get `There are no valid entry points for the dev server. Be sure to set the ViteHelper.development.scriptEntries config.` but as you said `This is currently only working in production mode`. Can you make it work for dev mode too? We're almost there!

**Comment by passchn** [comment]
Great that it works with your setup! It's not (completely) working with mine (with dynamic imports) though. 

The exception means what it says. You need to define your entries in the `app_vite.php` config. It has nothing todo technically with the `filter`. 

For now there is no code splitting during development. But the plugin needs to know your entry points.

**Comment by passchn** [comment]
The `filter` filters instances of `ManifestRecord`, but these don't exist in development. I can add another option for dev-entries though.. But the `filter` is for production i.e. when there is a manifest.

**Comment by t0byman** [comment]
> For now there is no code splitting during development. But the plugin needs to know your entry points.
That's fine I guess during dev but can't you just read all the entries from the vite.config build.rollupOptions.input then? Or is it hard to read out that config? Would make it very DRY though!

**Comment by t0byman** [comment]
Can't get
```
'development' => [
    scriptEntries' => ['/var/www/html/resources/scripts/main.ts']
]
```
 to work

**Comment by passchn** [comment]
Had the idea to look at the vite config from php, too. But it's a javascript object and I don't know how easy it would be to parse that. 

The scriptEntries must be relative to project root. So i guess it should start with `resources/`.

**Comment by t0byman** [comment]
```
'development' => [
    'scriptEntries' => ['resources/scripts/main.ts'],
    'styleEntries' => ['resources/styles/main.css'],
    'Url' => 'http://localhost:5173',
],
```
The paths work now but it's not respecting my Url setting anymore, maybe a bug somewhere? Also looking at `vendor\passchn\cakephp-vite\config\app_vite.php`, shouldn't HostNeedles and Url be `lowerCamelCase` just like all the other config settings are?

**Comment by passchn** [comment]
That was a bug. Both were due to the nesting of the config. Fixed in https://github.com/brandcom/cakephp-vite/commit/148e517a0685a4eedc03c4ffbbfbc2e66e3469b7 and https://github.com/brandcom/cakephp-vite/commit/a10993c84db0550fdc380b013ad0a7c0cb284db6

**Comment by passchn** [comment]
@t0byman do you have an idea about this one? 
https://github.com/brandcom/cakephp-vite/issues/8 

Not critical though.. One can now set `development.outDirectory` to `false` if it should be the webroot: 
https://github.com/brandcom/cakephp-vite/commit/e367e9d965d8267e0be0e9ee1803148add0106c7

**Comment by passchn** [comment]
I feel everything is solved here besides the new issue #9 that I created to me as a todo :)

**Comment by t0byman** [comment]
Discovered a new bug but I'll create a new issue for that
